```
   Compiling style_derive v0.0.1 (file:///home/servo/buildbot/slave/linux-dev/build/components/style_derive)
error[E0432]: unresolved import `syn::visit::Visitor`
  --> /home/servo/buildbot/slave/linux-dev/build/components/style_derive/to_css.rs:99:9
   |
99 |     use syn::visit::Visitor;
   |         ^^^^^^^^^^^^^^^^^^^ Could not find `visit` in `syn`

error[E0433]: failed to resolve. Could not find `visit` in `syn`
   --> /home/servo/buildbot/slave/linux-dev/build/components/style_derive/to_css.rs:113:13
    |
113 |             syn::visit::walk_path(self, path);
    |             ^^^^^^^^^^^^^^^^^^^^^ Could not find `visit` in `syn`

error: cannot continue compilation due to previous error
```

--- a/components/style_derive/Cargo.toml
+++ b/components/style_derive/Cargo.toml
@@ -11,5 +11,5 @@ proc-macro = true
 
 [dependencies]
 quote = "0.3"
-syn = "0.11"
+syn = { version = "0.11", features = ["visit"] }
 synstructure = "0.5.2"

--- a/components/style_derive/to_css.rs
+++ b/components/style_derive/to_css.rs
@@ -18,12 +18,16 @@ pub fn derive(input: syn::DeriveInput) -> quote::Tokens {
     let match_body = synstructure::each_variant(&input, &style, |bindings, variant| {
         let mut identifier = to_css_identifier(variant.ident.as_ref());
         let mut expr = if let Some((first, rest)) = bindings.split_first() {
-            where_clause.predicates.push(where_predicate(first.field.ty.clone()));
+            if has_free_params(&first.field.ty, &input.generics.ty_params) {
+                where_clause.predicates.push(where_predicate(first.field.ty.clone()));
+            }
             let mut expr = quote! {
                 ::style_traits::ToCss::to_css(#first, dest)
             };
             for binding in rest {
-                where_clause.predicates.push(where_predicate(binding.field.ty.clone()));
+                if has_free_params(&binding.field.ty, &input.generics.ty_params) {
+                    where_clause.predicates.push(where_predicate(binding.field.ty.clone()));
+                }
                 expr = quote! {
                     #expr?;
                     ::std::fmt::Write::write_str(dest, " ")?;
@@ -88,6 +92,31 @@ pub fn derive(input: syn::DeriveInput) -> quote::Tokens {
             }
         }
     }
+}
+
+/// Returns whether `ty` is parameterized by any parameter from `params`.
+fn has_free_params(ty: &syn::Ty, params: &[syn::TyParam]) -> bool {
+    use syn::visit::Visitor;
+
+    struct HasFreeParams<'a> {
+        params: &'a [syn::TyParam],
+        has_free: bool,
+    }
+
+    impl<'a> Visitor for HasFreeParams<'a> {
+        fn visit_path(&mut self, path: &syn::Path) {
+            if !path.global && path.segments.len() == 1 {
+                if self.params.iter().any(|param| param.ident == path.segments[0].ident) {
+                    self.has_free = true;
+                }
+            }
+            syn::visit::walk_path(self, path);
+        }
+    }
+
+    let mut visitor = HasFreeParams { params: params, has_free: false };
+    visitor.visit_ty(ty);
+    visitor.has_free
 }
 
 /// `#ty: ::style_traits::ToCss`

